# <span style="color:#FF00FF"> Предобработка данных </span>

Отработка алгоритма решения задачи предобработки данных. 

Датасет <code>dtp_last.csv</code> содержит информацию о ДТП в Российской Федерации в 2018 году.

<li>'reg_code' - код региона (октмо)
<li>'регион' - название региона
<li>'road_code' - код дороги
<li>'Дорога' -  название дороги
<li>'road_type' - тип дороги
<li>'address' -  адрес
<li>'crash_type_name' -тип ДТП
<li>'crash_date' - дата ДТП
<li>'crash_time' - время ДТП
<li>'crash reason' - причина ДТП
<li>'fatalities_amount' - число погибших
<li>'victims_amount' - число пострадавших
<li>'vehicles_amount' - число транспортных средств
<li>'participants_amount' - число участников



Задача: Подготовить данные для дальнейшего исследования.

Работу сопроводите выводами и комментариями к коду.

## <span style="color:#FF00FF">Навигация</span>
* [Изучение общей информации.](#1)
* [Осуществление предобработки заголовков.](#2)
* [Заполнение пропусков.](#3)
* [Нахождение проблем в данных.](#4)
* [Преобразование типов данных. Добавление столбцов datetime из имеющихся столбцов, используя разделитель T. Преобразование к типу данных "дата/время".](#5)
* [Удаление дубликатов.](#6)
* [Добавление столбцов с днями недели на русском языке.](#7)
* [Экспорт результатов в excel.](#8)
* [Общий вывод по работе.](#9)

### <a class="anchor" id="1"></a><span style="color:#FF00FF"> Загрузка данных. Общая информация </span>

In [310]:
import pandas as pd # импортируем библотеку pandas

In [311]:
df = pd.read_excel('dtp_last.xlsx') # загружаем библиотеку

In [312]:
df.head() # выводим первые пять строк для дальнейшего анализа собранной информации

,reg_code,регион,road_code,Дорога,road_type,address,crash_type_name,crash_date,crash_time,crash reason,fatalities_amount,victims_amount,vehicles_amount,participants_amount
0,46,Московская область,NaN,мкр. Авиационный-Лямцино-ММК,Региональная или межмуниципальная дорога общег...,"ЦФО, Московская область, Городской округ Домод...",Опрокидывание,20181231,23:55:00,Другие нарушения ПДД водителями,1,0,1,12
1,53,Оренбургская область,53К-2118000,Обход г. Оренбурга,Региональная или межмуниципальная дорога общег...,"ПФО, Оренбургская область, Оренбург",Наезд на препятствие,20181231,23:55:00,Другие нарушения ПДД водителями,1,0,1,2
2,46,Московская область,А-107,"""Московское малое кольцо"" Икша - Ногинск - Бро...",Федеральная дорога общего пользования,"ЦФО, Московская область, Дмитровский район",Опрокидывание,20181231,23:40:00,Превышение установленной скорости движения,1,0,1,2
3,8,Хабаровский край,А-375,"""Восток"" Хабаровск - Красный Яр - Ариадное - Ч...",Федеральная дорога общего пользования,"ДФО, Хабаровский край, Хабаровский район",Наезд на пешехода,20181231,23:20:00,Нахождение на проезжей части без цели её перех...,0,1,1,2
4,35,Республика Крым,35Н-076,Симферополь - Севастополь - Бахчисарай,Региональная или межмуниципальная дорога общег...,"ЮФО, Республика Крым, Бахчисарайский район",Наезд на пешехода,20181231,23:00:00,Нарушение правил проезда пешеходного перехода,1,0,1,2


In [313]:
df.shape # размер датасета (для оценки объема данных)

(39102, 14)

In [314]:
df.info() # общая информация

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39102 entries, 0 to 39101
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   reg_code             39102 non-null  int64 
 1   регион               39083 non-null  object
 2   road_code            20777 non-null  object
 3   Дорога               38910 non-null  object
 4   road_type            38910 non-null  object
 5   address              39102 non-null  object
 6   crash_type_name      39102 non-null  object
 7   crash_date           39102 non-null  int64 
 8   crash_time           39102 non-null  object
 9   crash reason         38467 non-null  object
 10  fatalities_amount    39102 non-null  int64 
 11  victims_amount       39102 non-null  int64 
 12  vehicles_amount      39102 non-null  int64 
 13  participants_amount  39102 non-null  object
dtypes: int64(5), object(9)
memory usage: 4.2+ MB


#### <span style="color:lightblue"> <span style="color:deepskyblue">Вывод</span> </span>

1. Датасет dtp_last.csv содержит информацию о ДТП в Российской Федерации в 2018 году.
2. Размер данных датасета: 39102 строки и 14 столбцов (данных достаточно для статистической значимости, кол-во признаков (14 столбцов) также достаточно для анализа).
3. Далее проведем анализ типа данных и количества непустых значений:
   - первый (1) столбец "reg_code" имеет целочисленный тип и 39102 заполненных значений (нет пустых значений);
   - второй (2) столбец "регион" имеет строковый тип данных и 39083 заполненных значений (есть пустые значения);
   - третий (3) столбец "road_code" имеет строковый тип данных и 20777 заполненных значений (есть пустые значения);
   - четвертый (4) столбец "Дорога" имеет строковый тип данных и 38910 заполненных значений (есть пустые значения);
   - пятый (5) столбец "road_type" имеет строковый тип данных и 38910 заполненных значений (есть пустые значения);
   - шестой (6) столбец "address" имеет строковый тип данных и 39102 заполненных значений (нет пустых значений);
   - седьмой (7) столбец "crash_type_name" имеет строковый тип данных и 39102 заполненных значений (нет пустых значений);
   - восьмой (8) столбец "crash_date" имеет целочисленный тип данных и 39102 заполненных значений (нет пустых значений);
   - девятый (9) столбец "crash_time" имеет строковый тип данных и 39102 заполненных значений (нет пустых значений);
   - десятый (10) столбец "crash reason" имеет строковый тип данных и 38467 заполненных значений (есть пустые значения);
   - одиннадцатый (11) столбец "fatalities_amount" имеет целочисленный тип данных и 39102 заполненных значений (нет пустых значений);
   - двенадцатый (12) столбец "victims_amount" имеет целочисленный тип данных и 39102 заполненных значений (нет пустых значений);
   - тринадцатый (13) столбец "vehicles_amount" имеет целочисленный тип данных и 39102 заполненных значений (нет пустых значений);
   - четырнадцатый (14) столбец "participants_amount" имеет строковый тип данных и 39102 заполненных значений (нет пустых значений).

#### <span style="color:deepskyblue">Таким образом, 5 столбцов с целочисленным типом данных, другие 9 имеют строковый тип данных, наибольшее кол-во пропущенных значений содержит столбец "road_code" (39102 - 20777 = 18325 пропущенных значений). </span>

 ### <a class="anchor" id="2"></a><span style="color:#FF00FF"> Предобработка заголовков: </span>

In [315]:
df.columns = ['reg_code', 'region', 'road_code', 'road', 'road_type', 'address', 'crash_type_name', 'crash_date', 'crash_time', 'crash_reason', 'fatalities_amount', 'victims_amount', 'vehicles_amount', 'participants_amount']
# переименуем заголовки (для единообразия и удобства)

In [316]:
df.head()

,reg_code,region,road_code,road,road_type,address,crash_type_name,crash_date,crash_time,crash_reason,fatalities_amount,victims_amount,vehicles_amount,participants_amount
0,46,Московская область,NaN,мкр. Авиационный-Лямцино-ММК,Региональная или межмуниципальная дорога общег...,"ЦФО, Московская область, Городской округ Домод...",Опрокидывание,20181231,23:55:00,Другие нарушения ПДД водителями,1,0,1,12
1,53,Оренбургская область,53К-2118000,Обход г. Оренбурга,Региональная или межмуниципальная дорога общег...,"ПФО, Оренбургская область, Оренбург",Наезд на препятствие,20181231,23:55:00,Другие нарушения ПДД водителями,1,0,1,2
2,46,Московская область,А-107,"""Московское малое кольцо"" Икша - Ногинск - Бро...",Федеральная дорога общего пользования,"ЦФО, Московская область, Дмитровский район",Опрокидывание,20181231,23:40:00,Превышение установленной скорости движения,1,0,1,2
3,8,Хабаровский край,А-375,"""Восток"" Хабаровск - Красный Яр - Ариадное - Ч...",Федеральная дорога общего пользования,"ДФО, Хабаровский край, Хабаровский район",Наезд на пешехода,20181231,23:20:00,Нахождение на проезжей части без цели её перех...,0,1,1,2
4,35,Республика Крым,35Н-076,Симферополь - Севастополь - Бахчисарай,Региональная или межмуниципальная дорога общег...,"ЮФО, Республика Крым, Бахчисарайский район",Наезд на пешехода,20181231,23:00:00,Нарушение правил проезда пешеходного перехода,1,0,1,2


#### <span style="color:lightblue"> <span style="color:deepskyblue">Вывод</span> </span>
1. Заголовки были приведены к единому формату (теперь стандартизированы) ;
2. Заголовки охватывают все этапы дорожно-транспортных происшествий (начиная от кода региона до точного времени и типа аварии), в дальнейшей легко выполним отчет с детализацией от причины аварии до конкретного адреса/региона.

## <a class="anchor" id="3"></a> <span style="color:#FF00FF"> Обработка пропущенных значений: </span>

In [317]:
df.isna().sum() # анализируем кол-во пропущенных значений в каждом из столбцов

reg_code                   0
region                    19
road_code              18325
road                     192
road_type                192
address                    0
crash_type_name            0
crash_date                 0
crash_time                 0
crash_reason             635
fatalities_amount          0
victims_amount             0
vehicles_amount            0
participants_amount        0
dtype: int64

#### <span style="color:lightblue"> <span style="color:deepskyblue">Вывод</span> </span>
1. Первое, что бросается в глаза, огромное количество пропущенных значений в столбце "road_code" (18325/39101 = 0,469, т.е. приблизительно 47% данных - это пропущенные значения), использовать группировку данных по коду дороги не рекомендуется.
2. "crash_reason": для 635 дорожно-транспортных происшествий причина их возникновения не известна (1,6% данных этого столбца - неизвестные значения, приемлемо).
3. "road" и "road_type" имеют по 192 пропуска, что составляет менее 0,5% от всего объема данных и не является критическим (скорее всего, пропуски в этих столбцах связаны между собой).
4. В столбце "region" всего 19 пустых значений (легко заполнить).
5. Все остальные столбцы: "reg_code" "address", "crash_type_name", "crash_date", "crash_time", "fatalities_amount", "victims_amount", "vehicles_amount", "participants_amount" не имеют пропущенных значений.
Таким образом, большой пробел в данных по коду дороги чуть ограничит анализ по этому признаку, но не мешает исследованию по другим показателям.

In [318]:
df[df['region'].isna()] # анализируем пропущенные значения в столбце "region"

,reg_code,region,road_code,road,road_type,address,crash_type_name,crash_date,crash_time,crash_reason,fatalities_amount,victims_amount,vehicles_amount,participants_amount
36583,78,NaN,NaN,ПРЕЧИСТОЕ-КОЗА-СЕМЕНОВСКОЕ,Региональная или межмуниципальная дорога общег...,"ЦФО, Ярославская область, Первомайский район",Опрокидывание,20180427,22:10:00,"Превышение установленной скорости движения, На...",1,1,1,2
36685,78,NaN,М-8,Подъезд к городу Кострома от а/д М-8,Федеральная дорога общего пользования,"ЦФО, Ярославская область, Ярославский район",Столкновение,20180427,3:50:00,Неправильный выбор дистанции,1,0,6,4
36725,78,NaN,NaN,ЯРОСЛАВЛЬ-КРАСНЫЕ ТКАЧИ-ШОПША,Региональная или межмуниципальная дорога общег...,"ЦФО, Ярославская область, Ярославский район",Столкновение,20180426,16:05:00,Неправильный выбор дистанции,1,0,2,3
36746,78,NaN,NaN,КРАСНЫЙ ПРОФИНТЕРН-ВЯТСКОЕ,Региональная или межмуниципальная дорога общег...,"ЦФО, Ярославская область, Некрасовский район",Наезд на пешехода,20180426,12:20:00,Переход через проезжую часть вне пешеходного п...,3,0,1,3
36808,78,NaN,NaN,ЛЮБИМ-ТРОИЦА,Региональная или межмуниципальная дорога общег...,"ЦФО, Ярославская область, Любимский район",Наезд на стоящее ТС,20180425,20:50:00,Нарушение правил расположения ТС на проезжей ч...,1,0,2,12
36908,78,NaN,NaN,ЯРОСЛАВЛЬ-РЫБИНСК,Региональная или межмуниципальная дорога общег...,"ЦФО, Ярославская область, Ярославский район, Н...",Столкновение,20180424,20:11:00,Выезд на полосу встречного движения,5,0,2,5
37146,78,NaN,М-8,"""Холмогоры"" Москва - Ярославль - Вологда – Арх...",Федеральная дорога общего пользования,"ЦФО, Ярославская область, Первомайский район",Столкновение,20180422,1:50:00,Выезд на полосу встречного движения,1,1,3,2
37164,78,NaN,М-8,"""Холмогоры"" Москва - Ярославль - Вологда – Арх...",Федеральная дорога общего пользования,"ЦФО, Ярославская область, Даниловский район",Наезд на пешехода,20180421,23:10:00,Нахождение на проезжей части без цели её перехода,1,0,1,3
37273,78,NaN,NaN,УГЛИЧ-НЕКОУЗ-БРЕЙТОВО,Региональная или межмуниципальная дорога общег...,"ЦФО, Ярославская область, Угличский район",Съезд с дороги,20180421,2:20:00,Нарушение правил расположения ТС на проезжей ч...,2,0,1,2
37350,78,NaN,NaN,"ОБХОД ЯРОСЛАВЛЯ С МОСТОМ ЧЕРЕЗ р,ВОЛГА",Региональная или межмуниципальная дорога общег...,"ЦФО, Ярославская область, Ярославль, Дзержинский",Столкновение,20180420,9:40:00,Неправильный выбор дистанции,1,0,2,2


In [319]:
df['region'] = df['region'].fillna('Ярославская область') 
# заполняем пропуски в столбце "region" значением "Ярославская область", так как все 19 пробелов имеют подобный адресс

In [320]:
df.isna().sum() # осуществляем проверку заполнения пропущенных значений в столбце "region"

reg_code                   0
region                     0
road_code              18325
road                     192
road_type                192
address                    0
crash_type_name            0
crash_date                 0
crash_time                 0
crash_reason             635
fatalities_amount          0
victims_amount             0
vehicles_amount            0
participants_amount        0
dtype: int64

In [321]:
df[df['road_code'].isna()] # по аналогии 

,reg_code,region,road_code,road,road_type,address,crash_type_name,crash_date,crash_time,crash_reason,fatalities_amount,victims_amount,vehicles_amount,participants_amount
0,46,Московская область,NaN,мкр. Авиационный-Лямцино-ММК,Региональная или межмуниципальная дорога общег...,"ЦФО, Московская область, Городской округ Домод...",Опрокидывание,20181231,23:55:00,Другие нарушения ПДД водителями,1,0,1,12
6,1,Алтайский край,NaN,Вторая Каменка - Гилево,Региональная или межмуниципальная дорога общег...,"СФО, Алтайский край, Локтевский район, Гилевск...",Наезд на пешехода,20181231,21:40:00,Несоответствие скорости конкретным условиям дв...,1,0,1,2
8,56,Пензенская область,NaN,"""Автодорога """"р.п. Беково - с. Варварино - р.п...",Региональная или межмуниципальная дорога общег...,"ПФО, Пензенская область, Бековский район, Верт...",Наезд на пешехода,20181231,21:20:00,Переход через проезжую часть в неустановленном...,1,0,1,2
12,7,Ставропольский край,NaN,Ачикулак - Махмуд-Мектеб Тукуй Мектеб 0000 - 4...,Региональная или межмуниципальная дорога общег...,"СКФО, Ставропольский край, Нефтекумский район,...",Опрокидывание,20181231,21:00:00,Другие нарушения ПДД водителями,4,0,1,4
13,3,Краснодарский край,NaN,г.Армавир-с.Успенское-г.Невинномысск,Региональная или межмуниципальная дорога общег...,"ЮФО, Краснодарский край, Успенский район, Коно...",Столкновение,20181231,20:59:00,Несоблюдение очередности проезда,1,0,2,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
39085,63,Саратовская область,NaN,"""автомобильная дорога """"Самара - Пугачев - Энг...",Региональная или межмуниципальная дорога общег...,"ПФО, Саратовская область, Энгельсский район",Столкновение,20180401,0:05:00,Выезд на полосу встречного движения,2,0,3,2
39087,46,Московская область,NaN,МКАД-Крутицы-Железнодорожный-Ликино-Дулево,Региональная или межмуниципальная дорога общег...,"ЦФО, Московская область, Железнодорожный",Наезд на пешехода,20180401,4:15:00,Нахождение на проезжей части без цели её перехода,0,1,1,2
39095,46,Московская область,NaN,Куровское-Шатура-Дмитровский Погост-Самойлиха,Региональная или межмуниципальная дорога общег...,"ЦФО, Московская область, Орехово-Зуевский район",Столкновение,20180401,2:40:00,Несоблюдение очередности проезда,1,0,2,4
39098,79,Республика Адыгея,NaN,Подъезд к а.Вочепший,Региональная или межмуниципальная дорога общег...,"ЮФО, Республика Адыгея, Теучежский район",Наезд на животное,20180401,1:30:00,Другие нарушения ПДД водителями,0,1,1,12


In [322]:
df['road_code'] = df['road_code'].fillna('отсутствует') 
# из-за большого объема пропущенных значений (почти 50% данных) в ручную не удастся отсмотреть даже большей части пропусков, поэтому принимается решение заполнить все пробелы значением "отсутствует"

In [323]:
df.isna().sum()

reg_code                 0
region                   0
road_code                0
road                   192
road_type              192
address                  0
crash_type_name          0
crash_date               0
crash_time               0
crash_reason           635
fatalities_amount        0
victims_amount           0
vehicles_amount          0
participants_amount      0
dtype: int64

In [324]:
df[df['road'].isna()]

,reg_code,region,road_code,road,road_type,address,crash_type_name,crash_date,crash_time,crash_reason,fatalities_amount,victims_amount,vehicles_amount,participants_amount
229,52,Омская область,отсутствует,NaN,NaN,"СФО, Омская область, Знаменский район",Столкновение,20181230,15:00:00,Нарушение правил расположения ТС на проезжей ч...,0,1,3,3
347,87,Республика Коми,отсутствует,NaN,NaN,"СЗФО, Республика Коми, Усть-Вымский район",Опрокидывание,20181229,23:30:00,Несоответствие скорости конкретным условиям дв...,1,0,1,2
521,11,Архангельская область,отсутствует,NaN,NaN,"СЗФО, Архангельская область, Каргопольский район",Наезд на пешехода,20181229,5:10:00,Неожиданный выход из-за сооружений (деревьев),1,0,1,2
549,22,Нижегородская область,отсутствует,NaN,NaN,"ПФО, Нижегородская область, Краснобаковский ра...",Столкновение,20181228,22:15:00,"Превышение установленной скорости движения, Вы...",3,0,2,4
672,11,Архангельская область,отсутствует,NaN,NaN,"СЗФО, Архангельская область, Холмогорский район",Наезд на препятствие,20181228,12:00:00,Другие нарушения ПДД водителями,1,0,1,12
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
38158,66,Смоленская область,отсутствует,NaN,NaN,"ЦФО, Смоленская область, Велижский район",Наезд на пешехода,20180412,4:20:00,Другие нарушения ПДД водителями,1,0,1,2
38306,22,Нижегородская область,отсутствует,NaN,NaN,"ПФО, Нижегородская область, Воротынский район",Опрокидывание,20180410,11:30:00,Нарушение правил расположения ТС на проезжей ч...,1,0,1,12
38351,52,Омская область,отсутствует,NaN,NaN,"СФО, Омская область, Омский район",Столкновение,20180409,18:10:00,Несоблюдение очередности проезда,2,0,2,3
38838,22,Нижегородская область,отсутствует,NaN,NaN,"ПФО, Нижегородская область, Володарский район",Наезд на стоящее ТС,20180404,8:30:00,Несоответствие скорости конкретным условиям дв...,1,0,2,2


In [325]:
df[df['road_type'].isna()] 
# в ходе анализа пропущенных значений в столбцах "road" и "road_type" приходим к выводу, что эти столбцы имеют пропуски в одних и тех же строчках 

,reg_code,region,road_code,road,road_type,address,crash_type_name,crash_date,crash_time,crash_reason,fatalities_amount,victims_amount,vehicles_amount,participants_amount
229,52,Омская область,отсутствует,NaN,NaN,"СФО, Омская область, Знаменский район",Столкновение,20181230,15:00:00,Нарушение правил расположения ТС на проезжей ч...,0,1,3,3
347,87,Республика Коми,отсутствует,NaN,NaN,"СЗФО, Республика Коми, Усть-Вымский район",Опрокидывание,20181229,23:30:00,Несоответствие скорости конкретным условиям дв...,1,0,1,2
521,11,Архангельская область,отсутствует,NaN,NaN,"СЗФО, Архангельская область, Каргопольский район",Наезд на пешехода,20181229,5:10:00,Неожиданный выход из-за сооружений (деревьев),1,0,1,2
549,22,Нижегородская область,отсутствует,NaN,NaN,"ПФО, Нижегородская область, Краснобаковский ра...",Столкновение,20181228,22:15:00,"Превышение установленной скорости движения, Вы...",3,0,2,4
672,11,Архангельская область,отсутствует,NaN,NaN,"СЗФО, Архангельская область, Холмогорский район",Наезд на препятствие,20181228,12:00:00,Другие нарушения ПДД водителями,1,0,1,12
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
38158,66,Смоленская область,отсутствует,NaN,NaN,"ЦФО, Смоленская область, Велижский район",Наезд на пешехода,20180412,4:20:00,Другие нарушения ПДД водителями,1,0,1,2
38306,22,Нижегородская область,отсутствует,NaN,NaN,"ПФО, Нижегородская область, Воротынский район",Опрокидывание,20180410,11:30:00,Нарушение правил расположения ТС на проезжей ч...,1,0,1,12
38351,52,Омская область,отсутствует,NaN,NaN,"СФО, Омская область, Омский район",Столкновение,20180409,18:10:00,Несоблюдение очередности проезда,2,0,2,3
38838,22,Нижегородская область,отсутствует,NaN,NaN,"ПФО, Нижегородская область, Володарский район",Наезд на стоящее ТС,20180404,8:30:00,Несоответствие скорости конкретным условиям дв...,1,0,2,2


In [326]:
df['road'] = df['road'].fillna('не определяется')
df['road_type'] = df['road_type'].fillna('не определяется')
# из-за отсутствия информации о дорожном коде, самой дороге и ее типе принимается решение заполнить все пробелы значением "не определяется"

In [327]:
df.isna().sum()

reg_code                 0
region                   0
road_code                0
road                     0
road_type                0
address                  0
crash_type_name          0
crash_date               0
crash_time               0
crash_reason           635
fatalities_amount        0
victims_amount           0
vehicles_amount          0
participants_amount      0
dtype: int64

In [328]:
df[df['crash_reason'].isna()]

,reg_code,region,road_code,road,road_type,address,crash_type_name,crash_date,crash_time,crash_reason,fatalities_amount,victims_amount,vehicles_amount,participants_amount
56,718,Ханты-Мансийский автономный округ — Югра,отсутствует,а/д г.Сургут - г.Нижневартовск,Региональная или межмуниципальная дорога общег...,"УФО, Ханты-Мансийский автономный округ, Нижнев...",Опрокидывание,20181231,15:00:00,NaN,1,0,1,12
308,84,Республика Алтай,Р-256,"""Чуйский тракт"" Новосибирск - Барнаул - Горно-...",Федеральная дорога общего пользования,"СФО, Республика Алтай, Шебалинский район",Опрокидывание,20181230,9:00:00,NaN,1,0,2,2
375,80,Республика Башкортостан,отсутствует,"Арсланово-Кушнаренково(Дмитриевка,Бишкази,Ст.Т...",Региональная или межмуниципальная дорога общег...,"ПФО, Республика Башкортостан, Кушнаренковский ...",Наезд на пешехода,20181229,19:30:00,NaN,0,1,1,2
392,29,Калужская область,отсутствует,Калуга-Медынь,Региональная или межмуниципальная дорога общег...,"ЦФО, Калужская область, Дзержинский район",Столкновение,20181229,18:30:00,NaN,1,0,2,2
533,52,Омская область,Р-402,Тюмень - Ялуторовск - Ишим - Омск,Федеральная дорога общего пользования,"СФО, Омская область, Саргатский район",Опрокидывание,20181229,0:30:00,NaN,1,0,1,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
38765,45,Москва,А-109,Ильинское шоссе,Федеральная дорога общего пользования,"ЦФО, Москва, Участки обслуживания на территори...",Падение пассажира,20180405,11:10:00,NaN,1,0,1,2
38775,99,Еврейская автономная область,Р-297,"""Амур"" Чита - Невер - Свободный - Архара - Бир...",Федеральная дорога общего пользования,"ДФО, Еврейская автономная область, Смидовичски...",Опрокидывание,20180405,7:20:00,NaN,2,0,1,2
38946,11,Архангельская область,11К-612,Карпогоры - Сосновка - Нюхча - граница с Респу...,Региональная или межмуниципальная дорога общег...,"СЗФО, Архангельская область, Пинежский район",Столкновение,20180402,16:30:00,NaN,1,0,2,3
38950,35,Республика Крым,35К-013,Черноморское - Евпатория,Региональная или межмуниципальная дорога общег...,"ЮФО, Республика Крым, Черноморский район",Опрокидывание,20180402,15:05:00,NaN,1,0,1,2


In [329]:
df.head(3)

,reg_code,region,road_code,road,road_type,address,crash_type_name,crash_date,crash_time,crash_reason,fatalities_amount,victims_amount,vehicles_amount,participants_amount
0,46,Московская область,отсутствует,мкр. Авиационный-Лямцино-ММК,Региональная или межмуниципальная дорога общег...,"ЦФО, Московская область, Городской округ Домод...",Опрокидывание,20181231,23:55:00,Другие нарушения ПДД водителями,1,0,1,12
1,53,Оренбургская область,53К-2118000,Обход г. Оренбурга,Региональная или межмуниципальная дорога общег...,"ПФО, Оренбургская область, Оренбург",Наезд на препятствие,20181231,23:55:00,Другие нарушения ПДД водителями,1,0,1,2
2,46,Московская область,А-107,"""Московское малое кольцо"" Икша - Ногинск - Бро...",Федеральная дорога общего пользования,"ЦФО, Московская область, Дмитровский район",Опрокидывание,20181231,23:40:00,Превышение установленной скорости движения,1,0,1,2


In [330]:
df['crash_reason'] = df['crash_reason'].fillna('не известна')
# сначала возникла идея заполнения пропусков на основе данных из другого столбца и строк (кратко: анализируем "crash_type_name" нужной нам строки и определяем по другой строке с таким же "crash_type_name" причину аварии, но данная идея не во всех случаях истинна, потому что "crash_type_name" может быть схож, а причина аварии - иная)
# пример выше (0 и 1 строчка имеют "crash_type_name" = опрокидование, а "crash reason" отличаются)
# поэтому пропуски заполнены значением "не известна"

In [331]:
df.isna().sum()

reg_code               0
region                 0
road_code              0
road                   0
road_type              0
address                0
crash_type_name        0
crash_date             0
crash_time             0
crash_reason           0
fatalities_amount      0
victims_amount         0
vehicles_amount        0
participants_amount    0
dtype: int64

#### <span style="color:lightblue"> <span style="color:deepskyblue">Вывод</span> </span>
Таким образом, удалось избавиться от пустых значений во всех столбцах, данные готовы для дальнейшего анализа. Выводы теперь не будут искажены пропусками и будут более точными.

### <a class="anchor" id="4"></a> <span style="color:#FF00FF"> Поиск проблем в данных </span> 

In [332]:
df['reg_code'].unique() # проверяю коды регионов

array([ 46,  53,   8,  35,  26,   1,  80,  56,  79,   7,   3,  93,  57,
        97,  30,  25,  50,  29,  33,  99,  11,  70,  17,  22,  12,  24,
        78,  94,  19,   4,  60,  85, 718,  20,  71,  47,  41,  91,  28,
        88,  89,   5,  81,  87,  66,  32,  10,  54,  27,  75,  63,  86,
        83,  45,  61,  90,  82,  49,  37,  36,  40,  92,  76,  52,  65,
        15,  14,  68,  18,  64,  34,  67, 719,  42,  84,  44,  38,  69,
        95,  58,  73,  98,  96])

In [333]:
df['reg_code'].nunique() # проблем в этих данных нет, все "reg_code" соответствуют рассматриваемым регионам

83

In [334]:
df['region'].unique() # смотрим уникальные значения (уникальные регионы) для поиска проблем в данных

array(['Московская область', 'Оренбургская область', 'Хабаровский край',
       'Республика Крым', 'Республика Ингушетия', 'Алтайский край',
       'Республика Башкортостан', 'Пензенская область',
       'Республика Адыгея', 'Ставропольский край', 'Краснодарский край',
       'Республика Тыва', 'Пермский край',
       'Чувашская Республика — Чувашия', 'Камчатский край',
       'Иркутская область', 'Новосибирская область', 'Калужская область',
       'Кировская область', 'Еврейская автономная область',
       'Архангельская область', 'Тульская область',
       'Владимирская область', 'Нижегородская область',
       'Астраханская область', 'Ивановская область',
       'Ярославская область', 'Удмуртская Республика',
       'Вологодская область', 'Красноярский край', 'Ростовская область',
       'Республика Калмыкия', 'Ханты-Мансийский автономный округ — Югра',
       'Воронежская область', 'Тюменская область', 'Мурманская область',
       'Ленинградская область', 'Карачаево-Черкесская Рес

In [335]:
df['region'].nunique() # мы смотрим кол-во уникальных значений, чтобы оценить, если ли неточности в данных, ориентируясь на кол-во регионов в районе 84-89 штук

119

In [336]:
df['region'] = df['region'].str.replace(' обл.', ' область').str.replace('Лен.обл.', 'Ленинградская область') 
# заменяем, так как многие регионы имеют повторы типа "Вологодская область" и "Вологодская обл.", что явялется одним и тем же значением 

In [337]:
df['region'].nunique() 
# кол-во регинонов значительно уменьшилось, проблем в данных этого столбца больше не наблюдается

83

In [338]:
df['region'].unique() # еще раз проверяем неточности среди регионов

array(['Московская область', 'Оренбургская область', 'Хабаровский край',
       'Республика Крым', 'Республика Ингушетия', 'Алтайский край',
       'Республика Башкортостан', 'Пензенская область',
       'Республика Адыгея', 'Ставропольский край', 'Краснодарский край',
       'Республика Тыва', 'Пермский край',
       'Чувашская Республика — Чувашия', 'Камчатский край',
       'Иркутская область', 'Новосибирская область', 'Калужская область',
       'Кировская область', 'Еврейская автономная область',
       'Архангельская область', 'Тульская область',
       'Владимирская область', 'Нижегородская область',
       'Астраханская область', 'Ивановская область',
       'Ярославская область', 'Удмуртская Республика',
       'Вологодская область', 'Красноярский край', 'Ростовская область',
       'Республика Калмыкия', 'Ханты-Мансийский автономный округ — Югра',
       'Воронежская область', 'Тюменская область', 'Мурманская область',
       'Ленинградская область', 'Карачаево-Черкесская Рес

#### <span style="color:lightblue"> <span style="color:deepskyblue">Вывод</span> </span>
Практически полный охват территории страны (83 субъекта). Сейчас в России 89 субъектов, остальные данные либо не учитываются, либо приведена устаревшая информация (еще вариант - на территориях, которых нет в датасете, за указанный промежуток времени не произошло ни одного ДТП).

In [339]:
df['road_code'].unique() 

array(['отсутствует', '53К-2118000', 'А-107', ..., '16К-1097', '11К-612',
       '16К-1305'], dtype=object)

#### <span style="color:lightblue"> <span style="color:deepskyblue">Данные в порядке, это стандартная система кодирования дорог РФ.</span> </span>

In [340]:
df['road'].unique() # официальные названия автодорог

array(['мкр. Авиационный-Лямцино-ММК', 'Обход г. Оренбурга',
       '"Московское малое кольцо" Икша - Ногинск - Бронницы - Голицыно - Истра - Икша',
       ..., 'Самара - Бугуруслан - Угорье',
       'Чистополь - Аксубаево - Нурлат - Чувашское Енорускино',
       'Подъезд к а.Вочепший'], dtype=object)

In [341]:
df['road_type'].unique() # четкая классификация: федеральные/региональные или межмуниципальные/неопределенные

array(['Региональная или межмуниципальная дорога общего пользования',
       'Федеральная дорога общего пользования', 'не определяется'],
      dtype=object)

In [342]:
df['address'].unique() # единый формат: "округ, регион, район, населенный пункт"

array(['ЦФО, Московская область, Городской округ Домодедово',
       'ПФО, Оренбургская область, Оренбург',
       'ЦФО, Московская область, Дмитровский район', ...,
       'СЗФО, Псковская область, Новоржевский район, Вехнянская, д Литово',
       'ПФО, Пензенская область, Спасский район, Веденяпинский',
       'ПФО, Республика Башкортостан, Абзелиловский район, Равиловский, д Равилово'],
      dtype=object)

In [343]:
df['crash_type_name'].unique() # исправлен лишь один недочет ниже 

array(['Опрокидывание', 'Наезд на препятствие', 'Наезд на пешехода',
       'Столкновение', 'Наезд на стоящее ТС', 'Съезд с дороги',
       'Падение пассажира', 'Наезд на животное', 'Наезд на велосипедиста',
       'Иной вид ДТП', 'Падение груза',
       'Наезд на лицо, не являющееся участником дорожного движения(иного участника ДТП), осуществляющее несение службы',
       'Отбрасывание предмета(отсоединение колеса)',
       'Наезд на лицо, не являющееся участником дорожного движения(иного участника ДТП), осуществляющее производство работ',
       'Наезд на внезапно возникшее препятствие',
       'Наезд на гужевой транспорт',
       'Наезд на лицо, не являющееся участником дорожного движения(иного участника ДТП), осуществляющее какую-либо другую деятельность',
       'Возгорание вследствие технической неисправности движущегося или остановившегося ТС, участвующего в дорожном движении'],
      dtype=object)

In [344]:
df['crash_type_name'] = df['crash_type_name'].replace('Наезд на внезапно возникшее препятствие', 'Наезд на препятствие') # имеют общий смысл, поэтому было принято решение сделать подобную замену

In [345]:
df['crash_reason'].unique() # я решила оставить повторения, например,'Несоблюдение бокового интервала, Несоблюдение бокового интервала', потому что в ДТП часто участвует несколько водителей/пешеходов и т.д., для каждого из которых своя причина аварии, поэтому необходимо уточнение в данных

array(['Другие нарушения ПДД водителями',
       'Превышение установленной скорости движения',
       'Нахождение на проезжей части без цели её перехода, Другие нарушения ПДД водителями',
       'Нарушение правил проезда пешеходного перехода',
       'Выезд на полосу встречного движения',
       'Несоответствие скорости конкретным условиям движения',
       'Нарушение правил обгона',
       'Переход через проезжую часть в неустановленном месте (при наличии в зоне видимости перекрёстка), Другие нарушения ПДД водителями',
       'Другие нарушения ПДД водителями, Переход через проезжую часть вне пешеходного перехода в зоне его видимости либо при наличии в непосредственной близости подземного (надземного) пешеходного перехода',
       'Нарушение требований дорожных знаков',
       'Несоблюдение очередности проезда', 'Неправильный выбор дистанции',
       'Несоблюдение бокового интервала',
       'Нарушение правил расположения ТС на проезжей части',
       'Нахождение на проезжей части без 

In [346]:
df['fatalities_amount'].unique() 
# и в столбцах, что следуют далее, числовые данные, которые нет смысла рассматривать и исправлять (это реальные значения, а не ошибки)

array([ 1,  0,  2,  4,  3,  6,  7,  8,  5,  9, 14, 18, 10, 19, 13, 11, 27,
       17, 12, 24, 22, 15, 23, 20, 16, 46, 26])

#### <span style="color:deepskyblue"> Вывод </span>
1. Данные стали еще более качественными (избавились от проблем в датасете).
2. Анализ показал большой охват территории: 83 субъекта РФ из 89 существующих на данный момент и полный охват федеральных округов (ЦФО, ПФО, СЗФО и т.д.).
3. Просматривается единый формат адресов, проведена унификация названий регонов ("Вологодская обл." исправлено на "Вологодская область" для более точного дальнейшего анализа).

### <a class="anchor" id="5"></a> <span style="color:#FF00FF"> Преобразование типов данных </span>

In [347]:
df['date_time'] = df['crash_date'].astype(str) + 'T' + df['crash_time'].astype(str) 
# сначала мы объединяем два столбца с разделителем "Т" (требуется в задании), меняя тип данных на строковый (ранее тип данных столбца "crash_date" - целочисленный), так как сложение строк возможно только со строками;

In [348]:
df.head() # проверяем добавление нового столбца "date_time" в датасет

,reg_code,region,road_code,road,road_type,address,crash_type_name,crash_date,crash_time,crash_reason,fatalities_amount,victims_amount,vehicles_amount,participants_amount,date_time
0,46,Московская область,отсутствует,мкр. Авиационный-Лямцино-ММК,Региональная или межмуниципальная дорога общег...,"ЦФО, Московская область, Городской округ Домод...",Опрокидывание,20181231,23:55:00,Другие нарушения ПДД водителями,1,0,1,12,20181231T23:55:00
1,53,Оренбургская область,53К-2118000,Обход г. Оренбурга,Региональная или межмуниципальная дорога общег...,"ПФО, Оренбургская область, Оренбург",Наезд на препятствие,20181231,23:55:00,Другие нарушения ПДД водителями,1,0,1,2,20181231T23:55:00
2,46,Московская область,А-107,"""Московское малое кольцо"" Икша - Ногинск - Бро...",Федеральная дорога общего пользования,"ЦФО, Московская область, Дмитровский район",Опрокидывание,20181231,23:40:00,Превышение установленной скорости движения,1,0,1,2,20181231T23:40:00
3,8,Хабаровский край,А-375,"""Восток"" Хабаровск - Красный Яр - Ариадное - Ч...",Федеральная дорога общего пользования,"ДФО, Хабаровский край, Хабаровский район",Наезд на пешехода,20181231,23:20:00,Нахождение на проезжей части без цели её перех...,0,1,1,2,20181231T23:20:00
4,35,Республика Крым,35Н-076,Симферополь - Севастополь - Бахчисарай,Региональная или межмуниципальная дорога общег...,"ЮФО, Республика Крым, Бахчисарайский район",Наезд на пешехода,20181231,23:00:00,Нарушение правил проезда пешеходного перехода,1,0,1,2,20181231T23:00:00


In [349]:
df['date_time'] = pd.to_datetime(df['date_time'], format='%Y%m%dT%H:%M:%S') # преобразует строковый тип данных в формат даты и времени

In [350]:
df.info() # проверяем, что тип данных "date_time" изменился на "datetime64[ns]"

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39102 entries, 0 to 39101
Data columns (total 15 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   reg_code             39102 non-null  int64         
 1   region               39102 non-null  object        
 2   road_code            39102 non-null  object        
 3   road                 39102 non-null  object        
 4   road_type            39102 non-null  object        
 5   address              39102 non-null  object        
 6   crash_type_name      39102 non-null  object        
 7   crash_date           39102 non-null  int64         
 8   crash_time           39102 non-null  object        
 9   crash_reason         39102 non-null  object        
 10  fatalities_amount    39102 non-null  int64         
 11  victims_amount       39102 non-null  int64         
 12  vehicles_amount      39102 non-null  int64         
 13  participants_amount  39102 non-

In [351]:
df['participants_amount'] =  pd.to_numeric(df['participants_amount'], errors='coerce') 
# преобразует строковый тип данных в числовой, добавляя "errors='coerce'" (т.к. данные смешаные, присутствуют как числа (1, 2, 3 и т.д.), так и "больше 25"), "errors='coerce'" меняет строковый тип данных на "NaN" 

In [352]:
df['participants_amount'] = df['participants_amount'].fillna(0) #заполняем 0 пропуски, сделанные принудительно выше

In [353]:
df['participants_amount'] = df['participants_amount'].astype('int') # преобразуем 

In [354]:
df.info() # осуществляем проверку замены типа дынных столбца 'participants_amount' с "object" на "int64"

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39102 entries, 0 to 39101
Data columns (total 15 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   reg_code             39102 non-null  int64         
 1   region               39102 non-null  object        
 2   road_code            39102 non-null  object        
 3   road                 39102 non-null  object        
 4   road_type            39102 non-null  object        
 5   address              39102 non-null  object        
 6   crash_type_name      39102 non-null  object        
 7   crash_date           39102 non-null  int64         
 8   crash_time           39102 non-null  object        
 9   crash_reason         39102 non-null  object        
 10  fatalities_amount    39102 non-null  int64         
 11  victims_amount       39102 non-null  int64         
 12  vehicles_amount      39102 non-null  int64         
 13  participants_amount  39102 non-

#### <span style="color:deepskyblue">Вывод</span>
1. Поработав над типами данных, мы преобразовали столбец "date_time" в datetime64[ns] (из строкового типа).
2. "participants_amount" преобразован в int64 (из object).
3. Теперь датасет состоит из след. типов данных:
   строковый формат: 8 столбцов;
   дата/время: 1 столбец;
   целочисленный тип данных: 6 столбцов.

### <a class="anchor" id="6"></a> <span style="color:#FF00FF"> Поиск дубликатов </span>

In [355]:
df.duplicated() # просматриваем явные дубликаты (метод  возвращает объект Series со значениями True/False)

0        False
1        False
2        False
3        False
4        False
         ...  
39097     True
39098     True
39099     True
39100     True
39101     True
Length: 39102, dtype: bool

In [356]:
df.duplicated().sum() # находим число явных дубликатов

np.int64(17)

In [357]:
df.loc[df.duplicated(keep=False)] # просматриваем все явные дубликаты перед удалением

,reg_code,region,road_code,road,road_type,address,crash_type_name,crash_date,crash_time,crash_reason,fatalities_amount,victims_amount,vehicles_amount,participants_amount,date_time
37226,22,Нижегородская область,22К-0016,Семенов - Ковернино,Региональная или межмуниципальная дорога общег...,"ПФО, Нижегородская область, Городской округ Се...",Столкновение,20180421,14:40:00,Неправильный выбор дистанции,1,0,3,4,2018-04-21 14:40:00
37227,22,Нижегородская область,22К-0016,Семенов - Ковернино,Региональная или межмуниципальная дорога общег...,"ПФО, Нижегородская область, Городской округ Се...",Столкновение,20180421,14:40:00,Неправильный выбор дистанции,1,0,3,4,2018-04-21 14:40:00
39070,61,Рязанская область,М-5,"""Урал"" Москва - Рязань - Пенза - Самара - Уфа ...",Федеральная дорога общего пользования,"ЦФО, Рязанская область, Шацкий район",Наезд на препятствие,20180401,5:20:00,Нарушение правил расположения ТС на проезжей ч...,8,0,1,8,2018-04-01 05:20:00
39071,46,Московская область,отсутствует,МКАД-Крутицы-Железнодорожный-Ликино-Дулево,Региональная или межмуниципальная дорога общег...,"ЦФО, Московская область, Железнодорожный",Наезд на пешехода,20180401,4:15:00,Нахождение на проезжей части без цели её перехода,0,1,1,2,2018-04-01 04:15:00
39072,61,Рязанская область,Р-22,"«Каспий» автомобильная дорога М-4 ""Дон"" - Тамб...",Федеральная дорога общего пользования,"ЦФО, Рязанская область, Михайловский район",Столкновение,20180401,4:00:00,"Выезд на полосу встречного движения в местах, ...",1,1,4,4,2018-04-01 04:00:00
39073,27,Калининградская область,27А-007,Северный обход г. Калининграда,Региональная или межмуниципальная дорога общег...,"СЗФО, Калининградская область, Гурьевский район",Наезд на пешехода,20180401,3:50:00,Ходьба вдоль проезжей части попутного направле...,1,0,1,2,2018-04-01 03:50:00
39074,32,Кемеровская область,Р-255,Подъезд к городу Томск от а/д Р-255,Федеральная дорога общего пользования,"СФО, Кемеровская область, Юргинский район",Столкновение,20180401,3:50:00,Выезд на полосу встречного движения,1,0,2,2,2018-04-01 03:50:00
39075,61,Рязанская область,Р-132,Калуга - Тула - Михайлов - Рязань,Федеральная дорога общего пользования,"ЦФО, Рязанская область, Рязань, Железнодорожный",Наезд на препятствие,20180401,3:30:00,Нарушение правил расположения ТС на проезжей ч...,1,0,3,3,2018-04-01 03:30:00
39076,86,Республика Карелия,А-121,"""Сортавала"" Санкт-Петербург - Сортавала - авто...",Федеральная дорога общего пользования,"СЗФО, Республика Карелия, Лахденпохский район,...",Наезд на пешехода,20180401,3:20:00,Иные нарушения ПДД пешеходом,0,1,1,2,2018-04-01 03:20:00
39077,56,Пензенская область,М-5,"""Урал"" Москва - Рязань - Пенза - Самара - Уфа ...",Федеральная дорога общего пользования,"ПФО, Пензенская область, Кузнецкий район, Евла...",Столкновение,20180401,3:10:00,Выезд на полосу встречного движения,1,0,3,2,2018-04-01 03:10:00


In [358]:
df.drop_duplicates(inplace=True) # удаляем дубликаты и осущетвляем перезапись датафрейма с помощью "inplace=True"

In [359]:
df = df.drop_duplicates().reset_index(drop=True) # возвращаем последовательность индексов после удаления дубликатов 

In [360]:
columns = ['region', 'road', 'crash_type_name', 'crash_reason', 'date_time', 'fatalities_amount', 'victims_amount'] 
# определяю те столбцы, по которым далее будут просматриваться дубликаты

In [361]:
df[columns].duplicated().sum() # кол-во дубликатов

np.int64(1)

In [362]:
df.loc[df[columns].duplicated(keep=False)] # проматриваем дубликаты 
# есть одно отличие в данных "participants_amount" (36944 ст. - 3, 36945 ст. - 2), скорее всего, ошибка при сборе информации, так как все кроме этого аспекта совпадает (причина аварии - столкновение, кол-во транспортных средств в обоих вариантах - 2 штуки))

,reg_code,region,road_code,road,road_type,address,crash_type_name,crash_date,crash_time,crash_reason,fatalities_amount,victims_amount,vehicles_amount,participants_amount,date_time
36944,99,Еврейская автономная область,отсутствует,Биробиджан-Головино,Региональная или межмуниципальная дорога общег...,"ДФО, Еврейская автономная область, Биробиджанс...",Столкновение,20180424,11:45:00,Превышение установленной скорости движения,1,0,2,3,2018-04-24 11:45:00
36945,99,Еврейская автономная область,отсутствует,Биробиджан-Головино,Региональная или межмуниципальная дорога общег...,"ДФО, Еврейская автономная область, Биробиджанс...",Столкновение,20180424,11:45:00,Превышение установленной скорости движения,1,0,2,2,2018-04-24 11:45:00


In [363]:
data = df.drop_duplicates(subset=columns, keep='last') # удаляем дубликаты за исключением последнего вхождения

In [364]:
data[columns].duplicated().sum() # проверяем, что дубликаты удалены

np.int64(0)

#### <span style="color:deepskyblue">В итоге</span>

- обнаружено 17 явных дубликатов;
- обнаружен 1 скрытый дубликат.

Удаление дубликатов несет в себе смысл улучшения статистической значимости данных, так как информация становится более точной (улучшенная статистика).

### <a class="anchor" id="7"></a> <span style="color:#FF00FF"> Добавление столбцов </span>

In [365]:
df['wd'] = df['date_time'].dt.weekday # добавляем столбец "wd" с днем недели, используя метод .dt.weekday, который возвращает день недели в виде числа (0 - понедельник)

In [366]:
df.loc[df['wd'] == 0, 'wd_name'] = 'понедельник'
df.loc[df['wd'] == 1, 'wd_name'] = 'вторник'
df.loc[df['wd'] == 2, 'wd_name'] = 'среда'
df.loc[df['wd'] == 3, 'wd_name'] = 'четверг'
df.loc[df['wd'] == 4, 'wd_name'] = 'пятница'
df.loc[df['wd'] == 5, 'wd_name'] = 'суббота'
df.loc[df['wd'] == 6, 'wd_name'] = 'воскресенье'
# создаем еще один столбец, который каждому значению из столбца "wd" будет сопоставлять день недели (0 = "понедельник"); в задании требуется создать столбец с днями недели на русском языке

In [367]:
df.tail()

,reg_code,region,road_code,road,road_type,address,crash_type_name,crash_date,crash_time,crash_reason,fatalities_amount,victims_amount,vehicles_amount,participants_amount,date_time,wd,wd_name
39080,80,Республика Башкортостан,80Н-069,Аскарово - Альмухаметово - Сибай,Региональная или межмуниципальная дорога общег...,"ПФО, Республика Башкортостан, Абзелиловский ра...",Наезд на животное,20180401,1:55:00,Нарушение правил расположения ТС на проезжей ч...,2,0,1,4,2018-04-01 01:55:00,6,воскресенье
39081,79,Республика Адыгея,отсутствует,Подъезд к а.Вочепший,Региональная или межмуниципальная дорога общег...,"ЮФО, Республика Адыгея, Теучежский район",Наезд на животное,20180401,1:30:00,Другие нарушения ПДД водителями,0,1,1,12,2018-04-01 01:30:00,6,воскресенье
39082,3,Краснодарский край,Р-217,Подъезд к городу Майкоп от а/д Р-217,Федеральная дорога общего пользования,"ЮФО, Краснодарский край, Новокубанский район",Столкновение,20180401,1:00:00,"Выезд на полосу встречного движения в местах, ...",2,0,2,3,2018-04-01 01:00:00,6,воскресенье
39083,76,Забайкальский край,А-350,Чита - Забайкальск - граница с Китайской Народ...,Федеральная дорога общего пользования,"СФО, Забайкальский край, Читинский район, Ново...",Опрокидывание,20180401,0:30:00,Нарушение правил расположения ТС на проезжей ч...,3,1,1,4,2018-04-01 00:30:00,6,воскресенье
39084,63,Саратовская область,отсутствует,"""автомобильная дорога """"Самара - Пугачев - Энг...",Региональная или межмуниципальная дорога общег...,"ПФО, Саратовская область, Энгельсский район",Столкновение,20180401,0:05:00,Выезд на полосу встречного движения,2,0,3,2,2018-04-01 00:05:00,6,воскресенье


In [368]:
df['wd_name'].value_counts()

wd_name
суббота        6602
воскресенье    6522
пятница        5902
понедельник    5269
вторник        4958
четверг        4946
среда          4886
Name: count, dtype: int64

#### <span style="color:deepskyblue">Вывод</span>
На основе данных о днях недели можно сказать, что наиболее опасным днем для ДТП является суббота (6602 шт. - зафиксировано случаев в субботу).
Затем следует воскресенье (6522 случаев).
Третье место занимает пятница - 5902 ДТП.
Суббота и воскресенье занимают 33,6% от всего кол-ва аварий за приведенный промежуток времени (в выходные дни происходит резкий рост трафика: поездки на дачи, по магазинам и т.д., рост алкогольного фактора и другие причины).
Пятница также имеет высокий показатель по кол-ву ДТП, что может быть связано с усталостью за неделю, что в конечном счете приводит к невнимательности за рулем и происшествиям на дороге).

### <a class="anchor" id="8"></a> <span style="color:#FF00FF"> Экспорт результатов </span>

In [369]:
df.to_excel('dtp_last_new.xlsx') # экспортируем результат в excel

### <a class="anchor" id="9"></a> <span style="color:#FF00FF"> Общий вывод </span>

Как уже было сказано ранее, данные охватывают большое кол-во регионов (83 субъекта России) - высокое качество датасета.
1. На этапе загрузки мы оценили структуру данных, просмотрели общую информацию о датасете, выполнили первичный анализ.
2. Далее была выполнена предобработка заголовков, заполнены пропуски, решены проблемы в датасете.
3. Затем осуществлены измения типов данных для более корректного анализа.
6. Поработали над дубликатами (всего было удалено 18 дубликатов).
7. Добавили столбец с днями недели и пришли к выводу, что наиболее опасным днем для ДТП является суббота.
8. Затем экспортировали данные в excel.